Update infer documentation to include desciptions of Bootstrap and (#77)
PumasAI/PumasDocs.jl: docs/src/analysis/inference.md

# Inference

## The `infer` Function

`infer(fpm::FittedPumasModel[, type]; level=0.95)`

> The `infer` function computes uncertainty estimates for the population parameter estimates of the `fpm` structure and returns a `FittedPumasModelInference` structure used for inference based on the fitted model `fpm`. The uncertainty estimates are displayed in the form of standard errors and confidence intervals but the internal representation depends on the `type` of uncertainty estimator used. The standard errors can be extracted from the `FittedPumasModelInference` structure with the `stderror` function. The default level used in the calculation of the confidence intervals is `0.95`.

Reviewer: I just realized that we don't have documentation on `stderr`. Should we add that?

### Robust standard errors

> The default `type` of uncertainty estimator is the robust (also called sandwich) covariance estimator of [White1982](@cite). Internally, this is represented as the (robust) covariance matrix which can also be computed with the `vcov` function directly from the `FittedPumasModel` structure. The method requires that the observed information matrix is (numerically) positive definite and invertible which is not always the case. Then it is not possible to compute the robust covariance estimate, `infer` returns a failed `FittedPumasModelInference`.

Reviewer: ```suggestion
The default `type` of uncertainty estimator is the robust (also called sandwich) covariance estimator of [White1982](@cite). Internally, this is represented as the (robust) covariance matrix which can also be computed with the `vcov` function directly from the `FittedPumasModel` structure. The method requires that the observed information matrix is (numerically) positive definite and invertible which is not always the case. When it is not possible to compute the robust covariance estimate, `infer` returns a failed `FittedPumasModelInference`.
```

> The robust covariance estimator is robust to some degree of model misspecification, but the estimator relies on asymptotic normality of the maximum likelihood estimator. While this is often a reasonable approximation for the parameters associated the the mean of the response variable, it is often a poor approximation for the variance parameters. As a consequence, it is not unlikely that confidence intervals for the variance parameters contain negative values when based on the the robust covariance matrix estimator. For that reason, the uncertainty can be assessed with two other types of estimators: `Bootstrap` and `SIR`.

Reviewer: ```suggestion
The robust covariance estimator is robust to some degree of model misspecification, but the estimator relies on asymptotic normality of the maximum likelihood estimator. While this is often a reasonable approximation for the parameters associated with the mean of the response variable, it is often a poor approximation for the variance parameters. As a consequence, it is not unlikely that confidence intervals for the variance parameters contain negative values when based on the the robust covariance matrix estimator. For that reason, the uncertainty can be assessed with two other types of estimators: `Bootstrap` and `SIR`.
```

Reviewer: >it is not unlikely that confidence intervals for the variance parameters contain negative values when based on the the robust covariance matrix estimator.

two things - _not unlikely_ seems like a double negative, is it easier to say _likely_? Also, I did not know that the reason for using bootstrap and SIR are because we may possibly find negative values of the CI (usually that means that the CI includes zero and the parameter is not significant). Just wanted to clarify

### Bootstrapped standard errors (`Pumas.Bootstrap`)

> The quality of the standard errors and confidence intervals can typically be improved by a bootstrap procedure. By passing the `Bootstrap` structure as the second argument, the uncertainty estimates will be based on a *paired bootstrap*, [efron1979bootstrap](@cite), possibly with stratification one or more variables, see below.

Reviewer: ```suggestion
The quality of the standard errors and confidence intervals can typically be improved by a bootstrap procedure. By passing the `Bootstrap` structure as the second argument, the uncertainty estimates will be based on a *paired bootstrap*, [efron1979bootstrap](@cite), possibly with stratification on one or more variables, see below.
```

The `Pumas.Bootstrap` structure constructor takes several optional arguments to specify the details of the bootstrap.

`Pumas.Bootstrap(;
    rng::AbstractRNG,
    samples::Integer,
    stratify_by::Union{Nothing,Symbol},
    ensemblealg::Union{EnsembleSerial,EnsembleThreads,EnsembleDistributed}])`

By default, the bootstrap uses the global RNG of Julia but it is possible to pass a custom `rng`. The default number of `samples` is `200`. It is possible to stratify on a discrete covariate by passing the name of the covariate as the `stratify_by` argument. The default is no stratification. Finally, it is possible to specify the mode of parallelism through the `ensemblealg` argument. The default is `EnsembleThreads`. When running Pumas in a distributed setting, it can be beneficial to specify `EnsembleThreads` when fitting the model and then `EnsembleDistributed` for the bootstrap. Thereby, the different bootstrap samples will be distributed across Julia processes and each associated model fitting will benefit from multi-threaded execution.

### Sampling importance resampling (`Pumas.SIR`)

> While the bootstrap usually improves the quality of the standard errors and confidence intervals, it can be computationally prohibitive since the requires rerunning a potentially costly fitting procedure many times. Furthermore, when few subjects are available, a non-parameteric procedure might not work well or at all.

Reviewer: ```suggestion
While the bootstrap usually improves the quality of the standard errors and confidence intervals, it can be computationally prohibitive since this requires rerunning a potentially costly fitting procedure many times. Furthermore, when a few subjects are available, a non-parametric procedure might not work well or at all.
```

As an alternative to the bootstrap procedures, [dosne2016improving](@cite) suggests the sampling importance resampling procedure for estimating the uncertainty. The proposal distribution for the procedure is based on robust covariance matrix and will therefore fail whenever the computation of the robust covariance matrix fails.

The procedure is selected by passing the Pumas.SIR` structure to `infer`.

`Pumas.SIR(rng::AbstractRNG, samples::Integer, resamples::Integer)`

> The `samples` argument specifies the number of samples to be drawn from the proposal distribution and the argument is mandatory. The arguments `rng` and `resamples` are optional and specifies the `rng` to use for the random draws as well as the number of resamples to draw from the proposal samples. The defaults the standard Julia RNG and `samples ÷ 3` respectively.

Reviewer: I thought SIR was parallelized by default. Should we way something about it?

# Inference

## `infer`

```
infer(fpm::FittedPumasModel[, type]; level=0.95)
```

The `infer` function computes uncertainty estimates for the population parameter estimates of the `fpm` structure and returns a `FittedPumasModelInference` structure used for inference based on the fitted model `fpm`. The uncertainty estimates are displayed in the form of standard errors and confidence intervals but the internal representation depends on the `type` of uncertainty estimator used. The standard errors can be extracted from the `FittedPumasModelInference` structure with the [stderror](@ref) function. The default level used in the calculation of the confidence intervals is `0.95`.

> ### Robust standard errors

Reviewer: `vcov` - let's open an issue to remind ourselves to add this as a function to be documented, and that it takes an `fpm`

The robust covariance estimator is robust to some degree of model misspecification, but the estimator relies on asymptotic normality of the maximum likelihood estimator. While this is often a reasonable approximation for the parameters associated with the mean of the response variable, it is often a poor approximation for the variance parameters. As a consequence, it is not unlikely that confidence intervals for the variance parameters contain negative values when based on the the robust covariance matrix estimator. For that reason, the uncertainty can be assessed with two other types of estimators: `Bootstrap` and `SIR`.

### Bootstrapped standard errors (`Pumas.Bootstrap`)

> The quality of the standard errors and confidence intervals can typically be improved by a bootstrap procedure. By passing the `Bootstrap` structure as the second argument, the uncertainty estimates will be based on a *paired bootstrap*, [efron1979bootstrap](@cite), possibly with stratification on one or more variables, see below.

Reviewer: ```suggestion
The quality of the standard errors and confidence intervals can typically be improved by a bootstrap procedure. By passing the `Pumas.Bootstrap` structure as the second argument, the uncertainty estimates will be based on a *paired bootstrap*, [efron1979bootstrap](@cite), possibly with stratification on one or more variables, see below.
```

should it not be `Pumas.Bootsrap()`?

### Sampling importance resampling (`Pumas.SIR`)

While the bootstrap usually improves the quality of the standard errors and confidence intervals, it can be computationally prohibitive since this requires rerunning a potentially costly fitting procedure many times. Furthermore, when a few subjects are available, a non-parametric procedure might not work well or at all.

As an alternative to the bootstrap procedures, [dosne2016improving](@cite) suggests the sampling importance resampling procedure for estimating the uncertainty. The proposal distribution for the procedure is based on robust covariance matrix and will therefore fail whenever the computation of the robust covariance matrix fails.

> The procedure is selected by passing the Pumas.SIR` structure to `infer`.

Reviewer: ```suggestion
The procedure is selected by passing the `Pumas.SIR` structure to `infer`.
```

`Pumas.SIR(rng::AbstractRNG, samples::Integer, resamples::Integer)`

> The `samples` argument specifies the number of samples to be drawn from the proposal distribution and the argument is mandatory. The arguments `rng` and `resamples` are optional and specifies the `rng` to use for the random draws as well as the number of resamples to draw from the proposal samples. The defaults the standard Julia RNG and `samples ÷ 3` respectively.

Reviewer: ```suggestion
The `samples` argument specifies the number of samples to be drawn from the proposal distribution and the argument is mandatory. The arguments `rng` and `resamples` are optional and specifies the `rng` to use for the random draws as well as the number of resamples to draw from the proposal samples. The defaults are the standard Julia RNG and `samples ÷ 3` respectively.
```
